relay/circuit-v2.md

```diff
@@ -0,0 +1,398 @@
+`stop` protocol stream to _A_, sending a `CONNECT` message.
+
+Peer _A_ then responds to the relay with a `STATUS:OK` message, which
+responds to _B_ with a `STATUS:OK` message in the open `hop` stream
+and then proceeds to bridge the two streams into a relayed connection.
+The relayed connection flows in the `hop` stream between the
+connection initiator and the relay and in the `stop` stream between
+the relay and the connection termination point.
+
+### Hop Protocol
+
+The Hop protocol governs interaction between clients and the relay;
+it uses the protocol ID `/libp2p/circuit/relay/0.2.0/hop`.
+
+There are two parts of the protocol:
+- reservation, by peers that wish to receive relay service
+- connection initiation, by peers that wish to connect to a peer through the relay.
+
+#### Reservation
+
+In order to make a reservation, a peer opens a connection to the relay and sends a `HopMessage` with
+`type = RESERVE`:
+
+```
+HopMessage {
+  type = RESERVE
+}
+```
+
+The relay responds with a `HopMessage` of `type = STATUS`, indicating whether the reservation has
+been accepted.
+
+If the reservation is accepted, then the message has the following form:
+```
+HopMessage {
+  type = STATUS
+  status = OK
+  reservation = Reservation {...}
+  limit = Limit {...}
+}
+```
+
+The `reservation` field provides information about the reservation itself; the struct has the following fields:
+```
+Reservation {
+   expire = ...
+   addrs = [...]
+   voucher = ...
+}
+```
+
+- the `expire` field contains the expiration time as a UTC UNIX time. The reservation becomes invalid after this time and it's the responsibility of the client to refresh.
+- the `addrs` field contains all the public relay addrs, including the peer ID of the relay node but not the
+  trailing `p2p-circuit` part; the client can use this list to construct its
+  own `p2p-circuit` relay addrs for advertising by encapsulating
+ `p2p-circuit/p2p/QmPeer` where `QmPeer` is its peer ID.
+- the `voucher` is the binary representation of the reservation voucher -- see [Reservation Vouchers](#reservation-vouchers) for details.
+
+The `limit` field, if present, provides information about the limits applied by the relay in relayed connection. When omitted, it indicates that the relay does not apply any limits.
+The struct has the following fields:
+```
+Limit {
+  duration = ...
+  data = ...
+```
+- the `duration` field indicates the maximum duration of a relayed connection in seconds; if 0, there is no limit applied.
+- the `data` field indicates the maximum number of bytes allowed to be transmitted in each direction; if 0 there is no limit applied.
+
+Note that the reservation remains valid until its expiration, as long
+as there is an active connection from the peer to the relay. If the
+peer disconnects, the reservation is no longer valid.
+
+If the reservation is rejected, the relay responds with a `HopMessage` of the form
+```
+Reservation {
+  type = STATUS
+  status = ...
+}
+```
```

Comment:
I think moving this up to after the accepted case makes this spec more readable.